### PR TITLE
Warn when a secret is captured by an `aws.lambda.CallbackFunction`

### LINKED ISSUE #1524: Warn when a secret is captured by a aws.lambda.CallbackFunction
In https://github.com/pulumi/pulumi-aws/pull/1282 we added the ability for CallbackFunctions which captured a secret value to turn the entire captured function body into a Pulumi secret so that the sensitive data would be encrypted in Pulumi state.  

However, the sensitive values in this case would still be captured into the text of the Lambda function visibile inside AWS.  Although this is standard and expected behaviour for Pulumi secrets generally, it is also an easy way to "accidentally" leak a secret into a context that may not be trusted.  

We should consider warning when this happens, and using the opportunity to suggest one of the alternatives, ideally SecretsManager, but even environment variables would be "better".

### PR DESCRIPTION
**Pull request by lukehoban**
Fixes #1524

### CODE REVIEW

File: sdk/nodejs/lambda/lambdaMixins.ts [hunk 1]
```diff
@@ -337,7 +337,17 @@ export class CallbackFunction<E, R> extends LambdaFunction {
             closure, serializedFileNameNoExtension, args.codePathOptions);
 
         const code = pulumi.output(new pulumi.asset.AssetArchive(codePaths));
-        (<any>code).isSecret = closure.then(c => c.containsSecrets);
+        (<any>code).isSecret = closure.then(c => {
+            if (c.containsSecrets) {
+                pulumi.log.warn(`A secret value was captured and serialized into the body of the Lambda Function '${name}'. ` +
+                                `This value will be stored as an encyrpted Pulumi secret, but may be available in plain text ` +
+                                `inside the AWS deployment package.  You can use 'pulumi.unsecret' to convert the value to a non-secret ` +
```
Comment (cnunciato) [inline sdk/nodejs/lambda/lambdaMixins.ts:344]:
```suggestion
                                `inside the AWS deployment package. You can use 'pulumi.unsecret' to convert the value to a non-secret ` +
```

File: sdk/nodejs/lambda/lambdaMixins.ts [hunk 2]
```diff
@@ -337,7 +337,17 @@ export class CallbackFunction<E, R> extends LambdaFunction {
...
+                                `value if this is not a sensitive value, or else use Secrets Manager or environment variables to pass the` +
```
Comment (cnunciato) [inline sdk/nodejs/lambda/lambdaMixins.ts:345]: We just need one space here, I think:

```suggestion
                                `value if this is not a sensitive value, or else use Secrets Manager or environment variables to pass the ` +
```

For reference here's how this renders out:

image

**Review by cnunciato — COMMENTED** [review]
Couple super minor things I noticed when trying this out.

**Review by cnunciato — APPROVED** [review]

**Comment by joeduffy** [comment]
My only comment on warnings/errors like this is that, when confronted with similar things in Pulumi, I always wish we had a stack trace. It's often very difficult to pinpoint where exactly the resource/etc was being created, especially since it can be buried in layers of libraries. Not sure if there's a "systemic" way to fix that but wanted to mention in case others agree.

**Comment by lukehoban** [comment]
> My only comment on warnings/errors like this is that, when confronted with similar things in Pulumi, I always wish we had a stack trace. It's often very difficult to pinpoint where exactly the resource/etc was being created, especially since it can be buried in layers of libraries. Not sure if there's a "systemic" way to fix that but wanted to mention in case others agree.

Agreed with this.  I've updated to at least log the warning against the impacted resource.  Ideally we'd also include a "stack trace" or more likely a line of code, for where the secret was captured - but that would require a quite significant new investment in the closure serialization implementation.  On balance, I'm inclined to take this change without doing that now - and consider it in the future if this is an area of feedback. But I'm open to holding off on making any change here for now, and waiting until we can invest in the new closure serialization machinery if anyone feels that is preferable.

File: sdk/nodejs/lambda/lambdaMixins.ts [hunk 3]
```diff
@@ -337,7 +337,17 @@ export class CallbackFunction<E, R> extends LambdaFunction {
...
 
         const code = pulumi.output(new pulumi.asset.AssetArchive(codePaths));
-        (<any>code).isSecret = closure.then(c => c.containsSecrets);
+        (<any>code).isSecret = closure.then(c => {
+            if (c.containsSecrets) {
+                pulumi.log.warn(`A secret value was captured and serialized into the body of the Lambda Function '${name}'. ` +
+                                `This value will be stored as an encyrpted Pulumi secret, but may be available in plain text ` +
+                                `inside the AWS deployment package. You can use 'pulumi.unsecret' to convert the value to a non-secret ` +
+                                `value if this is not a sensitive value, or else use Secrets Manager or environment variables to pass the ` +
+                                `senstive data to your function.`, this);
```
Comment (cnunciato) [inline sdk/nodejs/lambda/lambdaMixins.ts:346]:
```suggestion
                                `sensitive data to your function.`, this);
```

File: sdk/nodejs/lambda/lambdaMixins.ts [hunk 4]
```diff
@@ -337,7 +337,17 @@ export class CallbackFunction<E, R> extends LambdaFunction {
             closure, serializedFileNameNoExtension, args.codePathOptions);
 
         const code = pulumi.output(new pulumi.asset.AssetArchive(codePaths));
-        (<any>code).isSecret = closure.then(c => c.containsSecrets);
+        (<any>code).isSecret = closure.then(c => {
+            if (c.containsSecrets) {
+                pulumi.log.warn(`A secret value was captured and serialized into the body of the Lambda Function '${name}'. ` +
+                                `This value will be stored as an encyrpted Pulumi secret, but may be available in plain text ` +
```
Comment (cnunciato) [inline sdk/nodejs/lambda/lambdaMixins.ts:343]:
```suggestion
                                `This value will be stored as an encrypted Pulumi secret, but may be available in plain text ` +
```

**Review by cnunciato — COMMENTED** [review]
Apologies for not noticing these earlier!

**Comment by cnunciato** [comment]
Are we thinking we'll merge this?

**Review by Zaid-Ajaj — APPROVED** [review]

**Pull request merged**